lib/pbench/server/api/resources/query_apis/datasets_publish.py

```diff
@@ -0,0 +1,207 @@
+        # this PR.
+
+        cmds = []
+        for index, ids in map.items():
+            for id in ids:
+                x = (
+                    f'{{"update": {{"_index": "{index}", "_id": "{id}"}}}}\n'
+                    f'{{"doc": {{"authorization": {{"access": "{access}"}}}}}}'
+                )
+                cmds.append(x)
+        data = "\n".join(cmds) + "\n"
+        return {
+            "path": "/_bulk",
+            "kwargs": {
+                "data": data,
+                "headers": {"Content-Type": "application/x-ndjson"},
+                "params": {"refresh": "true"},
+            },
+        }
+
+    def postprocess(
+        self, es_json: Dict[AnyStr, Any], context: Dataset
+    ) -> Dict[AnyStr, Any]:
+        """
+        A bulk update succeeds if any of the individual updates succeeds, and
+        PUBLISH can affect hundreds of documents across many Elasticsearch
+        indices. Here we'll look for the "errors" boolean in the response
+        payload, which is true if any failures occurred. We'll then scan the
+        "items" list to diagnose the specific failures in the log.
+
+        FIXME: What do we do about error recovery here? At this point we may
+        have "published" some subset of the Elasticsearch documents but not
+        others. Do we try to "unpublish" those failed documents and fail the
+        whole operation? (And what if some of the "unpublish" operations fail
+        as well?)
+
+        NOTE: The Elasticsearch response "item" looks like this, and we expect
+        each item to be an "update" operation. the "error" field will only
+        exist if the item's "status" isn't successful.
+
+                {
+                    "update": {
+                        "_index": "index1",
+                        "_type" : "_doc",
+                        "_id": "5",
+                        "status": 404,
+                        "error": {
+                            "type": "document_missing_exception",
+                            "reason": "[_doc][5]: document missing",
+                            "index_uuid": "aAsFqTI0Tc2W0LCWgPNrOA",
+                            "shard": "0",
+                            "index": "index1"
+                        }
+                    }
+                },
+        """
+        msg = ""
+        status = HTTPStatus.OK
+        if es_json["errors"]:
+            for i in es_json["items"]:
+                u = i["update"]
+                if u["status"] != HTTPStatus.OK:
+                    e = u["error"]
+                    self.logger.error(
+                        "{} ({}: {}) for id {} in index {}",
+                        u["status"],
+                        e["type"],
+                        e["reason"],
+                        u["_id"],
+                        u["_index"],
+                    )
+                    msg = "One or more Elasticsearch PUBLISH operations failed; see server log for details."
```

Comment:
I don't want to return an `abort` message with hundreds of document IDs, either. (And, yeah, in which case the log is going to get blasted with a lot of noise, but I don't want to lose the state and I don't see a great alternative.)

Reply:
I basically OK with dumping everything into the log.  What I don't like is suggesting that the end user consult a log of which s/he has no knowledge or to which s/he has no access.

Perhaps we could return the ES reason from the first (or last) failure?  That seems like it might be better than a generic "ES failed you" message.

Reply:
I'm considering that still, though I'm not entirely convinced, partly because I'm not exactly sure which status codes Elasticsearch might return, or how to choose one that's representative and doesn't imply a user error where there actually is none. Still, I don't like using an internal error (although at least internal errors are generally considered potentially re-try-able, and 4xx errors generally aren't).

Reply:
Yeah...if we had a closed list of all the ES errors, that would make classifying them much more straightforward.  As for choosing, I only suggested choosing one because of the simplicity of implementation.  It's actually pretty easy, using a Python `dict`, to capture all of the ES errors while counting and discarding duplicates -- the net result is likely to be quite small (usually when bulk ingestions or updates fail, nearly all the documents fail for the same reason) -- so it wouldn't be hard to report back what we got.  And, given that, reporting it under either `INTERNAL_ERROR` or `GATEWAY_ERROR` would be fine, because the message text could convey the fact that it was one or more ES errors (and here they are...).